File: src/main/java/br/com/blz/testjava/dao/ProductDao.java
package br.com.blz.testjava.dao;

import br.com.blz.testjava.domain.Product;
import br.com.blz.testjava.exception.DuplicationException;
import br.com.blz.testjava.exception.NotFoundException;

public interface ProductDao {
 
    void save(Product product) throws DuplicationException;

    void update(Product product, Long sku) throws NotFoundException;

    Product findBySku(Long sku) throws NotFoundException;

    void delete(Long sku);
}

File: src/main/java/br/com/blz/testjava/ProductDao.kt
package br.com.blz.testjava

import br.com.blz.testjava.model.Product
import java.util.concurrent.ConcurrentHashMap

class ProductDao {

  private val storage = ConcurrentHashMap<Int, Product>()

  fun create(product: Product) {
    storage.putIfAbsent(product.sku, product)
  }

  fun update(product: Product) {
    storage.put(product.sku, product)
  }

  fun delete(sku: Int) {
    storage.remove(sku)
  }

  fun get(sku: Int): Product? {
    return storage[sku]
  }

}
